File: 01/elixir/solution.exs
defmodule Solution do
  def part_one(path) do
    file_content = File.read!(path)

    file_content
    |> String.trim()
    |> String.split(~r/\n/)
    |> Enum.map(&String.split/1)
    |> List.foldl([[], []], fn [x, y], [l1, l2] ->
      [[String.to_integer(x) | l1], [String.to_integer(y) | l2]]
    end)
    |> Enum.map(&Enum.sort/1)
    |> List.zip()
    |> Enum.map(fn {l, r} -> abs(l - r) end)
    |> Enum.sum()
    |> IO.puts()
  end

  def part_two(path) do
    [left, right] =
      path
      |> File.read!()
      |> String.trim()
      |> String.split(~r/\n/)
      |> Enum.map(&String.split/1)
      |> List.foldl([[], []], fn [x, y], [l1, l2] ->
        [[String.to_integer(x) | l1], [String.to_integer(y) | l2]]
      end)

    right_freqs = Enum.frequencies(right)

    left
    |> Enum.map(fn x -> x * Map.get(right_freqs, x, 0) end)
    |> Enum.sum()
    |> IO.puts()
  end
end

System.argv() |> Enum.at(0) |> Solution.part_two()

File: 01/python/solution.py
from collections import defaultdict
from heapq import heappop, heappush
from sys import argv


def part_one(path):
    left, right = [], []
    nb_elems = 0
    with open(path, "r") as f:
        for line in f:
            l, r = line.strip().split()
            heappush(left, int(l))
            heappush(right, int(r))
            nb_elems += 1
    sorted_left = (heappop(left) for _ in range(nb_elems))
    sorted_right = (heappop(right) for _ in range(nb_elems))
    res = sum(abs(l - r) for (l, r) in zip(sorted_left, sorted_right))
    print(res)


def part_two(path):
    left, right = defaultdict(int), defaultdict(int)
    with open(path, "r") as f:
        for line in f:
            l, r = line.strip().split()
            left[int(l)] += 1
            right[int(r)] += 1
    res = 0
    for l, count_left in left.items():
        count_right = right[l]
        res += l * count_left * count_right
    print(res)


if __name__ == "__main__":
    part_one(argv[1])
    part_two(argv[1])
